pallets/did/src/did_details.rs

```diff
@@ -399,59 +400,74 @@ impl<T: Config> DidDetails<T> {
 		&mut self,
 		new_authentication_key: DidVerificationKey,
 		block_number: BlockNumberOf<T>,
-	) {
+	) -> Result<(), StorageError> {
 		let old_authentication_key_id = self.authentication_key;
 		let new_authentication_key_id = utils::calculate_key_id::<T>(&new_authentication_key.clone().into());
 		self.authentication_key = new_authentication_key_id;
 		// Remove old key ID from public keys, if not used anymore.
 		self.remove_key_if_unused(&old_authentication_key_id);
 		// Add new key ID to public keys. If a key with the same ID is already present,
 		// the result is simply that the block number is updated.
-		self.public_keys.insert(
-			new_authentication_key_id,
-			DidPublicKeyDetails {
-				key: new_authentication_key.into(),
-				block_number,
-			},
-		);
+		self.public_keys
+			.try_insert(
+				new_authentication_key_id,
+				DidPublicKeyDetails {
+					key: new_authentication_key.into(),
+					block_number,
+				},
+			)
+			.map_err(|_| StorageError::MaxPublicKeysPerDidKeyIdentifierExceeded)?;
+		Ok(())
 	}
 
 	/// Add new key agreement keys to the DID.
 	///
 	/// The new keys are added to the set of verification keys.
 	pub fn add_key_agreement_keys(
 		&mut self,
-		new_key_agreement_keys: BTreeSet<DidEncryptionKey>,
+		new_key_agreement_keys: BoundedBTreeSet<DidEncryptionKey, T::MaxTotalKeyAgreementKeys>,
```

Comment:
New key agreement keys are bounded by a different config parameter.
```suggestion
		new_key_agreement_keys: BoundedBTreeSet<DidEncryptionKey, T::MaxNewKeyAgreementKeys>,
```